### ISSUE TITLE
Complete User testing of Persona

### ISSUE THREAD

**Issue by davidschober** [avalonmediasystem/avalon#3451]
## Description
After Persona is available in our dev environment, perform testing to ensure it meets user expectations. The tests should attempt to:
- login as another user and perform tasks using "become" as a super admin
- Ensure we can login using a local strategy
- Add users as a super admin
- remove users as a super admin
- reset passwords as a user
- reset passwords as an admin

## Done looks like
- tested
- issues written if necessary.

**Comment by davidschober** [comment]
- login as another user and perform tasks using "become" as a super admin
  + Was able to become a user, but could not click the "back to admin" button. Errored with "you do not have the rights to manage users" . I had been logged in as archivist1. 
- Ensure we can login using a local strategy
  + @cjcolvar can you verify
- Add users as a super admin
  + success
  + Was able to invite users
  + user could set password 
- remove users as a super admin
  + deleted a user worked. Trying to re-add the user with the same error surfaced a devise error. 
- reset passwords as a user
  + didn't test, @joncameron can you test? 
- reset passwords as an admin
  + attempted multiple times, could not reset user password and have user successfully log in

**Comment by mcwhitaker** [comment]
I was also not able to login after changing someone's password.

**Comment by cjcolvar** [comment]
- login as another user and perform tasks using "become" as a super admin
  + Worked as long as the user I was impersonating was also an administrator
- Ensure we can login using a local strategy
  + Login with Identity still works
  + Login with Email/Password as existing user created with Identity fails with "Invalid email or password."
  + I didn't test LTI or any other omniauth login (besides Identity).
- Add users as a super admin
  + Invitation email sent to new user and login succeeded.
- remove users as a super admin
  + deleted a user worked. Trying to re-add the user (via invitation or signing up with Email/Password) with the same name errored with a unique constraint exception. Trying to recreate user using Identity failed with "Email taken".
  + Couldn't quickly figure out how to remove user from database such that it could be re-added. (Removed from identity table but couldn't find in user table.)
- reset passwords as a user
- reset passwords as an admin
  + Was able to successfully use forgot password on Email/Password strategy for new invited Email/Password user.  Changing password as super admin also worked for this user.
  + Setting password as super admin for existing Identity user did not change Identity login password but allowed user to login with Email/Password strategy.

Other notes:
  + Layout of Email/Password strategy pages need small improvements.
  + Admin link in footer provides option of login with any configured strategies, but LTI doesn't make sense in that list because only an LMS can initiate it.  (And clicking the button leads to an error.)

**Comment by mcwhitaker** [comment]
@davidschober I think this has been fully tested now. Can we close the ticket? Will these problems be reported to Notch8 before our next meeting or during the meeting?

**Comment by davidschober** [comment]
This issue is public, so we can point Rob to it.

**Comment by davidschober** [comment]
sent to Rob

**Comment by orangewolf** [comment]
# To Fix #
- [ ] become feature not working if you try to become a non-admin user (as an admin user, when click become for a non-admin user, I should be logged in as the non-admin user.)


# Questions #
> Login with Email/Password as existing user created with Identity fails with "Invalid email or password."

This is what I would expect to happen. It was decided during one of our calls that users would either be SSO _or_ direct login. Identity strategy is treated like any other SSO provider. Therefor I would expect you to not be able to login via email/pass login to Identity users, unless they had a separate password added via the admin dashboard.

> Remove user as superadmin

When you delete a user they are archived / soft deleted. They would need to be restored or fully deleted. The scope `User.with_deleted` gets you those users. I think that viewing / restoring / destroying archived users would be a worthwhile enhancement, though I'm not sure its in scope for this run.

> Setting password as super admin for existing Identity user did not change Identity login password but allowed user to login with Email/Password strategy.

Same here as under login above.

> Layout of Email/Password strategy pages need small improvements

Are there some specific asks here or you just asking us to go through them all again and check them?

> Admin link in footer provides option of login with any configured strategies, but LTI doesn't make sense in that list because only an LMS can initiate it. (And clicking the button leads to an error.)

I don't understand this one, can you clarify for me?

**Comment by cjcolvar** [comment]
@orangewolf Thanks for the quick review!

I think it would be helpful for us if we setup a more realistic configuration on a testing server.  I think having Omniauth Identity still enabled was confusing testing (hence why I was trying to test all possible paths to compare/contrast with previous comments).

That being said, I think your assessment is correct that Email/Password login is working as expected and persona's features work fine on users created through the Email/Password auth strategy (with the exception of becoming non-admin users as noted above).  I think it is on us to determine how to enable/disable persona (probably through documentation) and/or how to avoid the confusion of having two auth systems and persona's user management features only operating on one of them (Email/Password).

**Comment by cjcolvar** [comment]
> > Layout of Email/Password strategy pages need small improvements
> 
> Are there some specific asks here or you just asking us to go through them all again and check them?
I think it would be good for a PO to weigh in on this, but I was thinking adding some buffer between the checkbox and "Remember me" on the login page and some more buffer between the "Send me reset password instructions" button and the "Log in " link below it.  (There may be some in the password reset page as well but I forget.)
> 
> > Admin link in footer provides option of login with any configured strategies, but LTI doesn't make sense in that list because only an LMS can initiate it. (And clicking the button leads to an error.)
> 
> I don't understand this one, can you clarify for me?

LTI isn't a user initiated authentication strategy from within Avalon.  Instead a user initiates it through a LMS like canvas or blackboard and that posts to an omniauth callback with a hash from the LMS used for login.  At some point we introduced a "hidden" flag for omniauth providers in the avalon config so LTI (or any other provider an institution wanted to hide) could be suppressed on a page like this.  So this might be as simple as using `Avalon::Authentication::VisibleProviders` (https://github.com/avalonmediasystem/avalon/blob/master/lib/avalon/authentication.rb#L37).

**Comment by mcwhitaker** [comment]
@davidschober @joncameron  Chris is asking for POs to weigh in ^^^

**Comment by davidschober** [comment]
@cjcolvar @orangewolf I agree that there should be more padding around the sections on the login pages/forgot password pages. Ideally it should have about the same as other login pages. 

<img width="622" alt="Screen Shot 2019-07-30 at 8 37 32 AM" src="https://user-images.githubusercontent.com/4803230/62134083-84ddff00-b2a5-11e9-92f5-4b1c19b6694d.png">
<img width="550" alt="Screen Shot 2019-07-30 at 8 37 05 AM" src="https://user-images.githubusercontent.com/4803230/62134084-84ddff00-b2a5-11e9-9a08-718801fa0ea5.png">

**Comment by davidschober** [comment]
Regarding the other issues noted, @cjcolvar I'm not sure how I can test/verify. Can you walk us through?

**Comment by davidschober** [comment]
@orangewolf The Avalon team got together, and I realized there were a few parts I didn't understand about how persona was working under the hood. Hopefully this is a punchlist we can work from:

- [ ] The identity provider should be listed in the "manage users" table. There's a column for it, but it is not being updated.
- [ ] Change / Edit should be greyed out, not visible if the user is not a local user (devise). We discussed this as a solution to managers trying to edit passwords/reset passwords when they can't.
- [ ]  Remove admin from the footer. My notes had us settling on documentation for those setting up Avalon. 
- [ ] @cjcolvar was investigating whether there's an existing screen to use instead of the screen that comes up when you click "admin". At a minimum LTI needs to be removed. 
- [ ] Add some spacing around the "forgot password" and "sign up"
